### PR TITLE
Refactor commonalities between two approaches

### PR DESCRIPTION
**Pull request by awgu**
**Overview:**
This refactors some commonalities between the two approaches to overlapping DDP with ZeRO. This also partially addresses this comment: https://github.com/pytorch/pytorch/pull/62157#discussion_r680292827

**Test Plan:**
```
gpurun4 python test/distributed/optim/test_zero_redundancy_optimizer.py
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62624 Refactor commonalities between two approaches**
* #62623 Add invariant check (bucket indices: 0, 1, ..., k-1)

Differential Revision: [D30058543](https://our.internmc.facebook.com/intern/diff/D30058543)

### CODE REVIEW

**Comment by andwgu** [comment]
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30058543).

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 1]
```diff
@@ -13,6 +13,106 @@
...
+    # Construct the `gradients` input for the local optimizer step, which
+    # expects `None` in a list position to indicate that the corresponding
+    # parameter should not be updated
+    num_local_optim_params = len(zero.optim.param_groups[0]["params"])
+    gradients: List[Optional[torch.Tensor]] = \
+        [_NO_PARAM_UPDATE for _ in range(num_local_optim_params)]
+    assert bucket_index in overlap_info.offsets, \
+        f"Bucket index {bucket_index} was not assigned to rank {rank}"
+    offset = overlap_info.offsets[bucket_index]
+    bucket_gradients = bucket.gradients()
```
Comment (mrshenli) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:48]: nit: bucket_gradients seems just used once, looks like we don't need to create a var for it?

File: torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py [hunk 2]
```diff
@@ -13,6 +13,106 @@
...
+            src=source_rank,
+            async_op=True
+        )
+    )
+
+def _collect_ddp_bucket_info(
+    bucket: dist.GradBucket,
+    zero: ZeroRedundancyOptimizer,
+    rank: int,
+    rank_to_update: int,
```
Comment (mrshenli) [inline torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:87]: `owner_rank`?

And also, would I be correct if I assume `owner_rank` is always the same as the `source_rank` argument in `_broadcast_bucket`, as the owner of those params should both update and broadcast. If yes, let's consolidate these two args to use the same name.

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks for improving code readability.

**Comment by andwgu** [comment]
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30058543).

**Pull request closed without merging**